lib/charms/mysql/v0/async_replication.py

```diff
@@ -0,0 +1,777 @@
+                return
+
+            self._charm.unit.status = BlockedStatus("Standalone read-only unit.")
+            # reset flag to allow instances rejoining the cluster
+            self._charm.unit_peer_data["member-state"] = "waiting"
+            del self._charm.unit_peer_data["unit-initialized"]
+            if self._charm.unit.is_leader():
+                self._charm.app.status = BlockedStatus("Recreate or rejoin cluster.")
+                logger.info(
+                    "\n\tThis is a replica cluster and will be dissolved.\n"
+                    "\tThe cluster can be recreated with the `recreate-cluster` action.\n"
+                    "\tAlternatively the cluster can be rejoined to the cluster set."
+                )
+                # reset the cluster node count flag
+                del self._charm.app_peer_data["units-added-to-cluster"]
+                # set flag to persist removed from cluster-set state
+                self._charm.app_peer_data["removed-from-cluster-set"] = "true"
+
+        elif self.role.cluster_role == "primary":
+            if self._charm.unit.is_leader():
+                # only leader units can remove replica clusters
+                remote_data = self.get_remote_relation_data(event.relation) or {}
+                if cluster_name := remote_data.get("cluster-name"):
+                    if self._charm._mysql.is_cluster_in_cluster_set(cluster_name):
+                        self._charm.unit.status = MaintenanceStatus("Removing replica cluster")
+                        logger.debug(f"Removing replica cluster {cluster_name}")
+
+                        # force removal when cluster is invalidated
+                        force = self._charm._mysql.get_replica_cluster_status(cluster_name) in [
+                            "invalidated",
+                            "unknown",
+                        ]
+
+                        self._charm._mysql.remove_replica_cluster(cluster_name, force=force)
+                        logger.debug(f"Replica cluster {cluster_name} removed")
+                        self._charm.unit.status = ActiveStatus(self._charm.active_status_message)
+                    else:
+                        logger.warning(
+                            f"Replica cluster {cluster_name} not found in cluster set, skipping removal"
+                        )
+                else:
+                    # Relation being broken before setup, e.g.: due to replica with user data
+                    logger.warning("No cluster name found, skipping removal")
+            elif self._charm.unit_peer_data.get("member-state") == "waiting":
+                # set member-state to allow status update
+                # needed for secondaries status update when removing due to replica with user data
+                self._charm.unit_peer_data["member-state"] = "unknown"
+            self._charm._on_update_status(None)
+
+    def _on_rejoin_cluster_action(self, event: ActionEvent) -> None:
```

Comment:
should this method be in the base class? it looks like it is only used in the primary subclass

Reply:
I'm declaring all actions on the base class and add the observers on a single one, to avoid call duplication. Don't fell strongly about where to put it